### PR TITLE
fix(workflow-controller): Fix output artifact duplicated in a fan-in topology with the use of retryStrategy. Fixes #4508

### LINKED ISSUE #4508: Output artifact duplicated in a fan-in topology with the use of retryStrategy
## Summary 

An output artifact is duplicated in a fan-in topology when a pre-fan-in step producing the artifact specifies `retryStrategy`; the duplication occurs regardless of whether there is an actual retry or not. I can work around this duplication behavior by deduping the result on my end. That said, if this is a bug, it'd be great if it could be addressed on the Argo side.

## Diagnostics

What Kubernetes provider are you using?
v1.18

What version of Argo Workflows are you running? 
v2.11.7

Reproduction steps are borrowed from [StackOverflow](https://stackoverflow.com/questions/60569353/dynamic-fan-in-in-argo-workflows) (and [this](https://github.com/argoproj/argo/issues/2393) enhancement request) with a minor addition that `retryStrategy` is added in the `write` step (the retry limit `3` does not matter for reproducibility and can be arbitrary).
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: loops-param-result-
spec:
  entrypoint: loop-param-result-example
  templates:
  - name: loop-param-result-example
    steps:
    - - name: generate
        template: gen-number-list
    - - name: write
        template: output-number
        arguments:
          parameters:
          - name: number
            value: "{{item}}"
        withParam: "{{steps.generate.outputs.result}}"
    - - name: fan-in
        template: fan-in
        arguments:
          parameters:
          - name: numbers
            value: "{{steps.write.outputs.parameters}}"

  - name: gen-number-list
    script:
      image: python:alpine3.6
      command: [python]
      source: |
        import json
        import sys
        json.dump([i for i in range(20, 31)], sys.stdout)

  - name: output-number
    inputs:
      parameters:
      - name: number
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo {{inputs.parameters.number}} > /tmp/number.txt"]
    outputs:
      parameters:
        - name: number
          valueFrom:
            path: /tmp/number.txt
    retryStrategy:
      limit: 3

  - name: fan-in
    inputs:
      parameters:
        - name: numbers
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo received {{inputs.parameters.numbers}}"]
```
The above `fan-in` step produces an output that looks like
```
[{"number":"20"},{"number":"20"},{"number":"21"},{"number":"21"},{"number":"22"},{"number":"22"},{"number":"23"},{"number":"23"},{"number":"24"},{"number":"24"},{"number":"25"},{"number":"25"},{"number":"26"},{"number":"26"},{"number":"27"},{"number":"27"},{"number":"28"},{"number":"28"},{"number":"29"},{"number":"29"},{"number":"30"},{"number":"30"}]
```
when we should expect each number to appear only once.

Also, the same issue can be reproducible with a DAG instead of steps:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: loops-param-result-
spec:
  entrypoint: loop-param-result-example
  templates:
  - name: loop-param-result-example
    dag:
      tasks:
      - name: generate
        template: gen-number-list
      - name: write
        depends: "generate.Succeeded"
        template: output-number
        arguments:
          parameters:
          - name: number
            value: "{{item}}"
        withParam: "{{tasks.generate.outputs.result}}"
      - name: fan-in
        depends: "write.Succeeded"
        template: fan-in
        arguments:
          parameters:
          - name: numbers
            value: "{{tasks.write.outputs.parameters}}"

  - name: gen-number-list
    script:
      image: python:alpine3.6
      command: [python]
      source: |
        import json
        import sys
        json.dump([i for i in range(20, 31)], sys.stdout)

  - name: output-number
    inputs:
      parameters:
      - name: number
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo {{inputs.parameters.number}} > /tmp/number.txt"]
    outputs:
      parameters:
        - name: number
          valueFrom:
            path: /tmp/number.txt
    retryStrategy:
      limit: 3

  - name: fan-in
    inputs:
      parameters:
        - name: numbers
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo received {{inputs.parameters.numbers}}"]
```
---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by smile-luobin**
Signed-off-by: smile-luobin <luobin_smile@163.com>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

Tips:

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it does not need to pass
* Sign-off your commits to pass the DCO check: `git commit --signoff`.
* Run `make pre-commit -B` to fix codegen or lint problems. 
* Say how how you tested your changes. If you changed the UI, attach screenshots.

### CODE REVIEW

**Comment by simster7** [comment]
@alexec Sure, I'll try to take a look today

**Review by simster7 — CHANGES_REQUESTED** [review]
Should we be using [`regexp.Compile`](https://pkg.go.dev/regexp#Compile) instead of `regexp.Match`? I may be mistaken, but if we use compile the regex string only once (which is an expensive op) instead of every time it is used.

**Comment by smile-luobin** [comment]
> Should we be using [`regexp.Compile`](https://pkg.go.dev/regexp#Compile) instead of `regexp.Match`? I may be mistaken, but if we use compile the regex string only once (which is an expensive op) instead of every time it is used.

fixed. this `regexp.Compile` is used to filter out grandchildNodes generated by childNodes with retryStrategy.

There is an another way to aggregate outputs for expanded tasks: 
1. use `strings.HasPrefix(node.Name, ancestorNode.Name+"(")` to collect `childNodes` (may be `childNodes` or `grandChildNodes`);
2. use `node.Type == PodNodeType` to collect the `podNodes` in `childNodes`;
3. do `processAggregateNodeOutputs` in all succeeded `podNodes` to get the aggregated outputs;

If this `podNodes` are `childNodes`, we can aggregate outputs in `podNodes` to get the aggregated outputs. 
If this `podNodes` are `grandChildNodes`, maybe we should to aggregate outputs in `childNodes`.  
But, the output of `childNode` is copied from the corresponding succeeded `grandChildNode`. So, we can also aggregate outputs in `podNodes`.

File: workflow/controller/dag.go [hunk 1]
```diff
@@ -498,10 +499,17 @@ func (woc *wfOperationCtx) buildLocalScopeFromTask(dagCtx *dagContext, task *wfv
 		}
 		prefix := fmt.Sprintf("tasks.%s", ancestor)
 		if ancestorNode.Type == wfv1.NodeTypeTaskGroup {
+			// This happens when there was `withItem/withParam` expansion.
+			// We add the aggregate outputs of our children to the scope as a JSON list
 			var ancestorNodes []wfv1.NodeStatus
+			retryNodeNamePattern, _ := regexp.Compile(`.*\([0-9]+\)$`)
```
Comment (simster7) [inline workflow/controller/dag.go:505]: This should be a `const` at the top of the file, so it's only called on once on program start up. Also, we can use `MustCompile`.

Something like

```go
const retryNodeNamePatter = regexp.MustCompile(`.*\([0-9]+\)$`)
```

At the top of this file is probably what we want. You should be able to find other examples of this in the codebase.

File: workflow/controller/steps.go [hunk 2]
```diff
@@ -112,9 +113,14 @@ func (woc *wfOperationCtx) executeSteps(ctx context.Context, nodeName string, tm
 				// This happens when there was `withItem/withParam` expansion.
 				// We add the aggregate outputs of our children to the scope as a JSON list
 				var childNodes []wfv1.NodeStatus
+				retryNodeNamePattern, _ := regexp.Compile(`.*\([0-9]+\)$`)
```
Comment (simster7) [inline workflow/controller/steps.go:116]: Once the previous comment is addressed, this can likely use the exam same variable as it is in the same package, no?

**Review by simster7 — CHANGES_REQUESTED** [review]
Sorry, I haven't had time to do a thorough review of this yet. But in my experience we should be very careful in extracting and inferring node types/relationships from their names. There can be a lot of edge cases. https://github.com/argoproj/argo-workflows/pull/5672 is an example of me dealing with this. In particular, take a look at the test cases on that PR. It is possible for user-defined output to be present on a node name, creating all sorts of edge cases when using regex or string manipulation to do these inferences, etc.

Again, I haven't had a chance to do a deep dive into this PR and confirm that this PR is safe for all edge cases (and even if I did, I wouldn't even be so confident with my own assertion anyway). Can you give me your thoughts on whether this is the most robust way to do this? Again, haven't done a deep dive here so I may be missing something.

File: workflow/controller/dag.go (hunk 1 above)
Comment (smile-luobin) [reply workflow/controller/dag.go:505]: > Sorry, I haven't had time to do a thorough review of this yet. But in my experience we should be very careful in extracting and inferring node types/relationships from their names. There can be a lot of edge cases. #5672 is an example of me dealing with this. In particular, take a look at the test cases on that PR. It is possible for user-defined output to be present on a node name, creating all sorts of edge cases when using regex or string manipulation to do these inferences, etc.
> 
> Again, I haven't had a chance to do a deep dive into this PR and confirm that this PR is safe for all edge cases (and even if I did, I wouldn't even be so confident with my own assertion anyway). Can you give me your thoughts on whether this is the most robust way to do this? Again, haven't done a deep dive here so I may be missing something.

Thanks. This is not the robust way to do this. And I found another bug about `Aggregated Outputs`: https://github.com/argoproj/argo-workflows/issues/6684. Currently, the outputs parameters of node with expansion expression is aggregated during processing childNode or grandChildNode. So the dag/steps outputs parameters cannot resolve these `aggregated outputs`. And i will these two issues in a new PR.

**Comment by smile-luobin** [comment]
> Sorry, I haven't had time to do a thorough review of this yet. But in my experience we should be very careful in extracting and inferring node types/relationships from their names. There can be a lot of edge cases. #5672 is an example of me dealing with this. In particular, take a look at the test cases on that PR. It is possible for user-defined output to be present on a node name, creating all sorts of edge cases when using regex or string manipulation to do these inferences, etc.
> 
> Again, I haven't had a chance to do a deep dive into this PR and confirm that this PR is safe for all edge cases (and even if I did, I wouldn't even be so confident with my own assertion anyway). Can you give me your thoughts on whether this is the most robust way to do this? Again, haven't done a deep dive here so I may be missing something.

Done in the latest patch. And I think it maybe the robust way to do this. As all node in wf.status.nodes have saved their childnodes ids in `Children` as a string list. So we can use this ids to get all these childNodes.

**Review by alexec — CHANGES_REQUESTED** [review]
Looks good to me. Please:

* [ ] Sync with master.
* [ ] Run `make pre-commit -B` and push changes.
* [ ] Dismiss reviews.

@simster7 might want to re-review?

**Comment by smile-luobin** [comment]
> Looks good to me. Please:
> 
> * [ ]  Sync with master.
> * [ ]  Run `make pre-commit -B` and push changes.
> * [ ]  Dismiss reviews.
> 
> @simster7 might want to re-revie

File: workflow/controller/dag.go [hunk 3]
```diff
@@ -515,10 +515,11 @@ func (woc *wfOperationCtx) buildLocalScopeFromTask(dagCtx *dagContext, task *wfv
 		}
 		prefix := fmt.Sprintf("tasks.%s", ancestor)
 		if ancestorNode.Type == wfv1.NodeTypeTaskGroup {
-			var ancestorNodes []wfv1.NodeStatus
-			for _, node := range woc.wf.Status.Nodes {
+			var childNodes = make([]wfv1.NodeStatus, len(ancestorNode.Children))
+			for i, nodeID := range ancestorNode.Children {
+				node := woc.wf.Status.Nodes[nodeID]
 				if node.BoundaryID == dagCtx.boundaryID && strings.HasPrefix(node.Name, ancestorNode.Name+"(") {
```
Comment (simster7) [inline workflow/controller/dag.go:521]: Is this check now moot if we're accessing the nodes directly from the children field?

File: workflow/controller/steps.go [hunk 4]
```diff
@@ -112,7 +112,8 @@ func (woc *wfOperationCtx) executeSteps(ctx context.Context, nodeName string, tm
 				// This happens when there was `withItem/withParam` expansion.
 				// We add the aggregate outputs of our children to the scope as a JSON list
 				var childNodes []wfv1.NodeStatus
-				for _, node := range woc.wf.Status.Nodes {
+				for _, childID := range sgNode.Children {
+					node := woc.wf.Status.Nodes[childID]
 					if node.BoundaryID == stepsCtx.boundaryID && strings.HasPrefix(node.Name, childNodeName+"(") && node.Type != wfv1.NodeTypeSkipped {
```
Comment (simster7) [inline workflow/controller/steps.go:117]: Similar with this check? Also, I wonder why the corresponding code in `dag.go` does not have `node.Type != wfv1.NodeTypeSkipped`?

**Review by simster7 — COMMENTED** [review]
Hey @smile-luobin, sorry for the crazy delay on this. The PR looks good to me, I just have a couple of minor questions. Also, if you could add some comments on the code to describe why we look at the children directly (i.e. referencing this issue), that would be great.

File: workflow/controller/dag.go (hunk 3 above)
Comment (smile-luobin) [reply workflow/controller/dag.go:521]: thanks. i think this check is no need any more.

File: workflow/controller/steps.go (hunk 4 above)
Comment (smile-luobin) [reply workflow/controller/steps.go:117]: Sorry, i need more time to review this code.
